### PR TITLE
modify CommonJS , Exceptions and BlockHeader of libethcore && modify other Files(modify sealegine and libevm/ExtVMFace to remove difficulty, author and DAO related logicals)

### PR DESCRIPTION
**Pull request by cyjseagull**
1. modify CommonJS: remove useless function `prettyU256`;
2.  modify Exceptions: mainly adjusting orders of exceptions according to their function;
3.  modify BlockHeader: remove difficulty, author, m_seal, and add m_sealer and m_sealerList
4. modify other Files to make compile success: mainly remove `difficulty, author and DAO` of `sealEngine, libevm/ExtVMFace`

### CODE REVIEW

File: libethcore/BlockHeader.h [hunk 1]
```diff
@@ -101,197 +77,186 @@ class BlockHeader
...
+    /// field 4: set logBloom
+    void setLogBloom(LogBloom const& _logBloom)
     {
-        m_author = _v;
+        m_logBloom = _logBloom;
         noteDirty();
     }
-    void setRoots(h256 const& _t, h256 const& _r, h256 const& _u, h256 const& _s)
+    /// field 5: set block number
+    void setNumber(int64_t _blockNumber)
```
Comment (wheatli) [inline libethcore/BlockHeader.h]: The _blockNumber should be u256

File: libethcore/BlockHeader.h [hunk 2]
```diff
@@ -101,197 +77,186 @@ class BlockHeader
...
-        m_number = _v;
+        m_gasUsed = _gasUsed;
         noteDirty();
     }
-    void setGasLimit(u256 const& _v)
+    /// field 8: set extra data
+    void setExtraData(bytes const& _content, unsigned int index = 0)
     {
-        m_gasLimit = _v;
+        m_extraData[index] = _content;
```
Comment (wheatli) [inline libethcore/BlockHeader.h]: You should check wether index is out of range here

File: libethcore/BlockHeader.h [hunk 3]
```diff
@@ -101,197 +77,186 @@ class BlockHeader
...
     h256 m_receiptsRoot;
     LogBloom m_logBloom;
     int64_t m_number = 0;
     u256 m_gasLimit;
     u256 m_gasUsed;
-    bytes m_extraData;
     int64_t m_timestamp = -1;
+    std::vector<bytes> m_extraData;  /// field for extension
+    /// Extended fields of FISCO-BCOS
+    u256 m_sealer = Invalid256;  // index of the sealer created this block
```
Comment (wheatli) [inline libethcore/BlockHeader.h]: good to be ///<

File: libethcore/BlockHeader.h (hunk 2 above)
Comment (wheatli) [reply libethcore/BlockHeader.h]: Maybe it's better to supply a method like appendExtraData which append bytes into m_extraData

File: libevm/ExtVMFace.cpp [hunk 4]
```diff
@@ -140,11 +140,11 @@ void getTxContext(evmc_tx_context* result, evmc_context* _context) noexcept
     auto& env = static_cast<ExtVMFace&>(*_context);
     result->tx_gas_price = toEvmC(env.gasPrice);
     result->tx_origin = toEvmC(env.origin);
-    result->block_coinbase = toEvmC(env.envInfo().author());
+    // result->block_coinbase = toEvmC(env.envInfo().author());
     result->block_number = static_cast<int64_t>(env.envInfo().number());
     result->block_timestamp = static_cast<int64_t>(env.envInfo().timestamp());
     result->block_gas_limit = static_cast<int64_t>(env.envInfo().gasLimit());
-    result->block_difficulty = toEvmC(env.envInfo().difficulty());
+    // result->block_difficulty = toEvmC(env.envInfo().difficulty());
```
Comment (wheatli) [inline libevm/ExtVMFace.cpp]: Advice: set these filed as a default value, and make a record that we should delete them is evmc lib

File: libethcore/BlockHeader.h [hunk 5]
```diff
@@ -101,197 +77,186 @@ class BlockHeader
...
 
+private:  /// private data fields
+    /// -------structure of block header------
     h256 m_parentHash;
-    h256 m_sha3Uncles;
     h256 m_stateRoot;
     h256 m_transactionsRoot;
     h256 m_receiptsRoot;
     LogBloom m_logBloom;
     int64_t m_number = 0;
```
Comment (wheatli) [inline libethcore/BlockHeader.h]: why not use u256?

File: libethcore/BlockHeader.h (hunk 1 above)
Comment (cyjseagull) [reply libethcore/BlockHeader.h]: has already been modified to u256.

File: libethcore/BlockHeader.h (hunk 2 above)
Comment (cyjseagull) [reply libethcore/BlockHeader.h]: has already modify this function, and add `appendExtraDataArray` and `appendExtraDataItem` functions.

please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/43/commits/64a3a5ff6f96964022d0f4ca933eaeb374947f9e

File: libethcore/BlockHeader.h (hunk 3 above)
Comment (cyjseagull) [reply libethcore/BlockHeader.h]: done

File: libevm/ExtVMFace.cpp (hunk 4 above)
Comment (cyjseagull) [reply libevm/ExtVMFace.cpp]: done, please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/43/commits/64a3a5ff6f96964022d0f4ca933eaeb374947f9e#diff-b2bd15a4ebfde03f3135170794bb9d19R116

File: libethcore/BlockHeader.h (hunk 5 above)
Comment (cyjseagull) [reply libethcore/BlockHeader.h]: has already been modified, please refer to https://github.com/FISCO-BCOS/lab-bcos/pull/43/commits/64a3a5ff6f96964022d0f4ca933eaeb374947f9e#diff-efee221d1491860184d229a955e08e4fR265

File: libethcore/BlockHeader.h [hunk 6]
```diff
@@ -173,19 +173,44 @@ class BlockHeader
...
         noteDirty();
     }
+
+    bool appendExtraDataItem(bytes const& _content, unsigned int _index = 0)
+    {
+        if (_index >= m_extraData.size())
+            return false;
+        else
+        {
+            for (uint64_t i = 0; i < _content.size(); i++)
```
Comment (wheatli) [inline libethcore/BlockHeader.h]: It's not a good idea...I think you can just append _content to m_extraData[_index]

Comment (cyjseagull) [reply libethcore/BlockHeader.h]: has been modified

File: libethcore/BlockHeader.h (hunk 1 above)
Comment (JimmyShi22) [reply libethcore/BlockHeader.h]: what's the difference between u256 and h256?

**Pull request merged**